media-proxy/include/mesh/st2110tx.h

```diff
@@ -0,0 +1,166 @@
+#ifndef ST2110TX_H
+#define ST2110TX_H
+
+#include "st2110.h"
+#include <algorithm>
+
+namespace mesh::connection {
+
+/**
+ * ST2110Tx
+ *
+ * Base abstract class of ST2110Tx. ST2110_20Tx/ST2110_22Tx/ST2110_30Tx
+ * inherit this class.
+ */
+template <typename FRAME, typename HANDLE, typename OPS> class ST2110Tx : public ST2110 {
+  public:
+    ST2110Tx() { _kind = Kind::transmitter; };
+    ~ST2110Tx() {
+        shutdown(_ctx);
+        if (ops.name)
+            free((void *)ops.name);
+    };
+
+  protected:
+    HANDLE mtl_session = nullptr;
+    OPS ops = {0};
+    uint32_t transfer_size = 0;
+    context::Context _ctx = context::WithCancel(context::Background());
+
+    virtual FRAME *get_frame(HANDLE) = 0;
+    virtual int put_frame(HANDLE, FRAME *) = 0;
+    virtual HANDLE create_session(mtl_handle, OPS *) = 0;
+    virtual int close_session(HANDLE) = 0;
+
+    int configure_common(context::Context& ctx, const std::string& dev_port,
+                         const MeshConfig_ST2110& cfg_st2110) {
+        int session_id = 0;
+        mtl_device = get_mtl_device(dev_port, MTL_LOG_LEVEL_CRIT, cfg_st2110.local_ip_addr, session_id);
+        if (!mtl_device) {
+            log::error("Failed to get MTL device");
+            return -1;
+        }
```

Comment:
Can you move this to the base class? You could make this `configure_common()` method virtual, override it in derived classes, and call the inherited method where needed.

Reply:
I suggest you extract as much as you can from `configure_common()` methods in derived classes to the base class method.

Reply:
Rework base class as template required